packages/utils/src/ast-utils/helpers.ts

```diff
@@ -1,75 +1,74 @@
 import { AST_NODE_TYPES, AST_TOKEN_TYPES, TSESTree } from '../ts-estree';
 
-type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntry<BaseType> = BaseType extends unknown
+  ? [keyof BaseType, BaseType[keyof BaseType]]
+  : never;
 type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
 
 export const isNodeOfType =
   <NodeType extends AST_NODE_TYPES>(nodeType: NodeType) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeType } =>
+  ): node is Extract<TSESTree.Node, { type: NodeType }> =>
     node?.type === nodeType;
 
 export const isNodeOfTypes =
   <NodeTypes extends readonly AST_NODE_TYPES[]>(nodeTypes: NodeTypes) =>
   (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeTypes[number] } =>
+  ): node is Extract<TSESTree.Node, { type: NodeTypes[number] }> =>
     !!node && nodeTypes.includes(node.type);
 
 export const isNodeOfTypeWithConditions = <
   NodeType extends AST_NODE_TYPES,
-  Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
+  ExtractedNode extends Extract<TSESTree.Node, { type: NodeType }>,
+  Conditions extends Partial<ExtractedNode>,
 >(
   nodeType: NodeType,
   conditions: Conditions,
 ): ((
   node: TSESTree.Node | null | undefined,
-) => node is TSESTree.Node & { type: NodeType } & Conditions) => {
+) => node is ExtractedNode & Conditions) => {
   const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Node>;
 
   return (
     node: TSESTree.Node | null | undefined,
-  ): node is TSESTree.Node & { type: NodeType } & Conditions =>
+  ): node is ExtractedNode & Conditions =>
     node?.type === nodeType &&
-    entries.every(([key, value]) => node[key] === value);
+    entries.every(([key, value]) => node[key as keyof TSESTree.Node] === value);
 };
 
 export const isTokenOfTypeWithConditions = <
   TokenType extends AST_TOKEN_TYPES,
+  ExtractedToken extends Extract<TSESTree.Token, { type: TokenType }>,
   Conditions extends Partial<TSESTree.Token & { type: TokenType }>,
 >(
   tokenType: TokenType,
   conditions: Conditions,
 ): ((
   token: TSESTree.Token | null | undefined,
-) => token is TSESTree.Token & { type: TokenType } & Conditions) => {
+) => token is ExtractedToken & Conditions) => {
   const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Token>;
 
   return (
     token: TSESTree.Token | null | undefined,
-  ): token is TSESTree.Token & { type: TokenType } & Conditions =>
+  ): token is ExtractedToken & Conditions =>
     token?.type === tokenType &&
-    entries.every(([key, value]) => token[key] === value);
+    entries.every(
+      ([key, value]) => token[key as keyof TSESTree.Token] === value,
+    );
 };
 
 export const isNotTokenOfTypeWithConditions =
   <
     TokenType extends AST_TOKEN_TYPES,
-    Conditions extends Partial<TSESTree.Token & { type: TokenType }>,
+    ExtractedToken extends Extract<TSESTree.Token, { type: TokenType }>,
```

Comment:
Mixed feelings about this. Certainly improves perf a lot, but it means we can't explicitly write the type parameters without a lot of duplication...

Reply:
Yeah, that's true.

Where explicit type parameters are used?

Reply:
Nowhere, just saying 😄

Reply:
I think the perf benefit is more important in this case, since we don't need to pass a type at all to the functions, as they are inferred by the passed value, right?